### PR TITLE
[mlir] [linalg] Add pattern to swap transpose with broadcast 

### PR DESCRIPTION
**Pull request by cxy-1993**
Add a pattern that implement:

  transpose(broadcast(input)) -> broadcast(transpose(input))

### CODE REVIEW

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
I understand this makes sense as something good to do always, but isnt this a one-off pattern that you want. Does it have to be a canonicalization where it needs a fixed point? (not hard-blocking though, cause I can see why this is better in general). (see the extremely long discussion here <https://discourse.llvm.org/t/rfc-update-to-general-design-section-of-operation-canonicalizations-in-mlir/79355>)

**Comment by cxy-1993** [comment]
> I understand this makes sense as something good to do always, but isnt this a one-off pattern that you want. Does it have to be a canonicalization where it needs a fixed point? (not hard-blocking though, cause I can see why this is better in general). (see the extremely long discussion here https://discourse.llvm.org/t/rfc-update-to-general-design-section-of-operation-canonicalizations-in-mlir/79355)

Thanks for the context. I completely agree with the discussion here: I also often struggle with whether an optimization pattern should be placed in fold/canonicalize or a pass. Maybe you should doc it here too(https://mlir.llvm.org/docs/Canonicalization/).

For this pattern, I think it's appropriate to use it as a canonicalization. Here are the reasons:
- This pattern is convergent.
- This pattern is unidirectional (in the direction of reducing data size, not like the case in the RFC).
- This is not a one-off pattern; new matches may be generated during the application process.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1917]: Unclear to me why there is a sort

**Review by MaheshRavishankar — COMMENTED** [review]
Thanks! Good to me for having as a canonicalization but under the caveat that if there is a valid case to move it out if canonicalization then we do that 
Note that such a move would cause downstream breakage. Something that was meant to happen automatically will change to it being opt in. This is another downside of canonicalization for me. Downstream uses can be unstable

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1916]: Nice catch! I must have mixed up the limitations of other ops and broadcasts, and mistakenly thought that the dimension of broadcasts must be an increasing sequence. The code has been modified, and an non-increasing order dimension test added.

**Comment by cxy-1993** [comment]
> Thanks! Good to me for having as a canonicalization but under the caveat that if there is a valid case to move it out if canonicalization then we do that Note that such a move would cause downstream breakage. Something that was meant to happen automatically will change to it being opt in. This is another downside of canonicalization for me. Downstream uses can be unstable

If this pattern breaks downstream repo, please feel free to remove it.

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (dcaballe) [inline mlir/include/mlir/Dialect/Utils/IndexingUtils.h:246]: remove -> removes?

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (dcaballe) [inline mlir/include/mlir/Dialect/Utils/IndexingUtils.h:250]: Is the result correct? Shouldn't this be `{2, 1, 3}` or am I getting this wrong?

File: mlir/lib/Dialect/Utils/IndexingUtils.cpp
Comment (dcaballe) [inline mlir/lib/Dialect/Utils/IndexingUtils.cpp:261]: nit: use pre-increment

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
Comment (dcaballe) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1895]: I wouldn't say "reduces the cost of transpose" as this may turn a transpose with a performant implementation into a simpler one that may or may not have a performant implementation. I would just describe this as a canonicalization pattern.

File: mlir/test/Dialect/Linalg/canonicalize.mlir
Comment (dcaballe) [inline mlir/test/Dialect/Linalg/canonicalize.mlir:1149]: Could you add a test with a `transpose(broadcast(1D->2D))` to see what happens?

**Review by dcaballe — COMMENTED** [review]
This looks like an interesting canonicalization, indeed! 

Two side effects we should watch out for: 

1) This may lead to not only lit test failures but also to performance regressions if we turn a transpose with a highly optimized lowering into a simpler transpose that falls into a default lowering path. 
2) We may get some performance improvements if we get rid of the transpose in `transpose(broadcast(1D->2D))` (assuming the transpose is not already optimized away somehow). We should add a test for this scenario.

An interesting follow-up would be to consider scenarios where a type conversion happens between the broadcast and the transpose.

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (cxy-1993) [reply mlir/include/mlir/Dialect/Utils/IndexingUtils.h:250]: Sorry not make this function clear.
This function returns a *new permutation* after removing  *input position* in removePositions.

The removed position is "2", "1" in input pos,  after remove, we have {4, 0, 3}.
To be a valid permutation, returned perm should start from "0", result should be {2, 0, 1}.

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (cxy-1993) [reply mlir/include/mlir/Dialect/Utils/IndexingUtils.h]: Done

File: mlir/lib/Dialect/Utils/IndexingUtils.cpp
Comment (cxy-1993) [reply mlir/lib/Dialect/Utils/IndexingUtils.cpp]: nice catch, done

File: mlir/test/Dialect/Linalg/canonicalize.mlir
Comment (cxy-1993) [reply mlir/test/Dialect/Linalg/canonicalize.mlir:1149]: DONE

**Comment by cxy-1993** [comment]
> This looks like an interesting canonicalization, indeed!
> 
> Two side effects we should watch out for:
> 
> 1. This may lead to not only lit test failures but also to performance regressions if we turn a transpose with a highly optimized lowering into a simpler transpose that falls into a default lowering path.
> 2. We may get some performance improvements if we get rid of the transpose in `transpose(broadcast(1D->2D))` (assuming the transpose is not already optimized away somehow). We should add a test for this scenario.
> 
> An interesting follow-up would be to consider scenarios where a type conversion happens between the broadcast and the transpose.

Thanks for the comment. I have considered the issue you mentioned. If there is a performance improvement for some special permutation transpose, then using expand shape should be a better solution. Of course, if this pattern breaks downstream repos, please feel free to discuss and remove it.

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (dcaballe) [reply mlir/include/mlir/Dialect/Utils/IndexingUtils.h:250]: Thanks! I think we should already have this functionality implemented on `AffineMap`. Would you mind taking a look at the utilities in AffineMap.h? There are some `drop...` methods might get you what you need. You can get an AffineMap from a permutation with: https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/AffineMap.h#L103

Comment (cxy-1993) [reply mlir/include/mlir/Dialect/Utils/IndexingUtils.h:250]: Thank you for your suggestion, this approach is very interesting (at least my function name is not precise enough, I think dropResults would be better). We can indeed replace the calculation of perm with the calculation of affine map using applyPermutationMap and getPermutationMap. Before making the changes, I would like to discuss some points: This will simplify the calculation of perm, but it will introduce more compilation time -- we have to construct affine maps to reuse the affine map operation functions. Is this a more reasonable approach? If so, all the util functions in the Permutation utils series have corresponding util functions in the affine map. Should we systematically replace them all?

**Comment by stellaraccident** [comment]
> Add canonicalize pattern that implement canonicalize:
> 
> transpose(broadcast(input)) -> broadcast(transpose(input))
> 
> Reduce the cost of transpose.

I'm not opposed to this being a canonicalization outright, but -- transpose and broadcast have no cost on their own/abstractly and justifying it based on that puts this in the realm of an optimization, not a canonicalization.

With that said, if we want to say that this is a canonical form, that  may be an ok thing to do, but we need to define it in terms of the lattice of all such ops. As Diego says above, there is often an interplay between several ops of this category and I'd prefer we not add a one off pattern without a better justification and analysis of what it means for all of them.

No matter what we do, let's not land such a canonicalization with a justification that it is an optimization (ie. Which is what the pr description states as the justification). I'm marking changes requested on that point and am happy to be overridden on that if the consensus is that this makes sense in terms of the lattice versus as a piece of a specific optimization strategy.

In general, there are entire optimization pipelines that are responsible for propagating various classes of layout and data movement operations. Such things are non trivial and are better done as an actual algorithm that can be controlled/cost modeled/etc.

**Review by stellaraccident — CHANGES_REQUESTED** [review]
As noted in my above comment, if there is a consensus that this is a canonicalization, please include that justification in the pr description vs justifying as an optimization. Marking change requested for that point. It seems to me that making this judgment would be part of a bit more analysis of adjacent ops vs a one off, but I won't stick on that point if others believe this is justified on its own.

(I do tend to agree this might be reasonable to do but the bar for calling something a canonicalization is high, and I'd like to see it justified and documented in such a way. Some of that commentary is presented inline and needs to be moved to the PR description and agreed on)

**Comment by cxy-1993** [comment]
> As noted in my above comment, if there is a consensus that this is a canonicalization, please include that justification in the pr description vs justifying as an optimization. Marking change requested for that point. It seems to me that making this judgment would be part of a bit more analysis of adjacent ops vs a one off, but I won't stick on that point if others believe this is justified on its own.
> 
> (I do tend to agree this might be reasonable to do but the bar for calling something a canonicalization is high, and I'd like to see it justified and documented in such a way. Some of that commentary is presented inline and needs to be moved to the PR description and agreed on)

Thank you for your suggestion. After receiving multiple concerns about placing this pattern in canonicalize, I have reconsidered the validity of this pattern in canonicalize. For some backends, doing so would indeed introduce unnecessary risks, so I have move this pattern from canonicalize to transform. Thanks again  @MaheshRavishankar @stellaraccident @dcaballe

**Comment by cxy-1993** [comment]
ping @MaheshRavishankar @stellaraccident :)

**Review by MaheshRavishankar — COMMENTED** [review]
Thanks! Looks reasonable to me. I wish we had a better grouping than a "swapTransposeWithBroadcast" kind of thing.. but I dont have a better idea.

I have no concerns here. Ill let @dcaballe finish his review of the implementation.

**Review by stellaraccident — COMMENTED** [review]
Lgtm

**Review by joker-eph — CHANGES_REQUESTED** [review]
> After receiving multiple concerns about placing this pattern in canonicalize, I have reconsidered the validity of this pattern in canonicalize

can we elaborate on this? I am concerned about a proliferation of random patterns exposed by various APIs that aren’t canonicalization when they could be: can we record these « concerns » as a rationale why this is not a good canonicalization?

**Comment by dcaballe** [comment]
Agree with Stella that this transformation, as a canonicalization, shouldn’t be driven by cost or performance considerations. The motivating factor should be to define a canonical order between these memory/data layout operations so that we can reduce the number of op combinations to be handled in other transformations. On top of that, transposing broadcasted data is transposing duplicated data so this transformation should make the IR less redundant, which is something that falls into the canonicalization territory, IMO. 

Defining a canonical order around memory/layout operations has been a long missing gap in our representation, resulting in downstreams adopting as “canonical” form whichever random order frontends or other transformations generated. This is suboptimal because these assumptions make our transformations fragile and we are more hesitant to accept changes that might alter these assumptions. 

Should this canonical representation be suitable for all the transformations? Not really. Each transformation may have a pre-processing step where the memory/layout operations can be reordered as needed. Even those pre-processing transformations would be simpler if they start from a canonical form.  

IMO, it would be hard to define and integrate a canonical form for the full lattice in one shot. However, this case looks like a no-brainer to me and it’s an incremental step towards addressing the current situation.

**Comment by MaheshRavishankar** [comment]
The real issue here is that this pattern is a one-off. It doesn't fit within anything else. It's not a pattern in service of a larger transformation goal. So appears like a zombie method.

**Comment by cxy-1993** [comment]
> > After receiving multiple concerns about placing this pattern in canonicalize, I have reconsidered the validity of this pattern in canonicalize
> 
> can we elaborate on this? I am concerned about a proliferation of random patterns exposed by various APIs that aren’t canonicalization when they could be: can we record these « concerns » as a rationale why this is not a good canonicalization?

I think this is a very good example to add to @MaheshRavishankar discussion of whether an optimization should use canonicalization. When we have a consensus, I will update the documentation to standardize future behavior.

My previous concern was that putting this optimization in canonicalization might not provide positive benefits for all backends. However, as dcaballe and stellaraccident mentioned, the goal of canonicalization should not be to achieve maximum performance on every backend. The goal of canonicalization is to make subsequent optimizations more effective. (This should be the original definition of canonicalization https://github.com/w3c/charmod-norm/blob/gh-pages/index.html, and we should at least agree on this point). 

According to this definition, if we define the direction of the lattice changed by canonicalization to reduce redundant data in the IR, I think this pattern is appropriate behavior for canonicalization, because as I mentioned in my previous comments, this pattern is convergent and unidirectional, and it is not a one-off pattern --- we may have other patterns that will generate new opportunities for this pattern.


> The real issue here is that this pattern is a one-off. It doesn't fit within anything else. It's not a pattern in service of a larger transformation goal. So appears like a zombie method.

For example, we can combine other data redundant reduce pattern to reach:
`reduce(transpose(broadcast(input))) - > broadcast(transpose(reduce(input)))`

Based on the discussion above, I believe that this pattern can be considered canonicalization, but we need to define the lattice and the direction of canonicalization carefully. Can we reach a consensus on this? @stellaraccident @MaheshRavishankar @dcaballe

**Comment by stellaraccident** [comment]
Thank you. That is the kind of justification I was looking for. Let's include that in the pr description and agree on it. 

(Sorry to add overhead to the process, but this area has gotten out of hand and we are trying to do better)

**Comment by cxy-1993** [comment]
I have updated the docs according the discuss above and https://discourse.llvm.org/t/rfc-update-to-general-design-section-of-operation-canonicalizations-in-mlir/79355. Please help me review it too. @MaheshRavishankar @stellaraccident @dcaballe @stellaraccident @joker-eph

**Comment by stellaraccident** [comment]
> I have updated the docs according the discuss above and https://discourse.llvm.org/t/rfc-update-to-general-design-section-of-operation-canonicalizations-in-mlir/79355. Please help me review it too. @MaheshRavishankar @stellaraccident @dcaballe @stellaraccident @joker-eph

Can you please update the description of this PR to include the rationale? Both you and Diego have a well enunciated view on that and we would like that in PR descriptions so it doesn't get lost in a centithread. Other than that lgtm.

**Comment by stellaraccident** [comment]
> Defining a canonical order around memory/layout operations has been a long missing gap in our representation, resulting in downstreams adopting as “canonical” form whichever random order frontends or other transformations generated. This is suboptimal because these assumptions make our transformations fragile and we are more hesitant to accept changes that might alter these assumptions.

This is well enunciated. Any chance once this patch lands that you could summarize it on discourse to leave breadcrumbs for those who come next. I think its fine to "incremental our way" to a better state, but we want to capture the thought process and plan a little more prominently. Good discussion on this thread on that front.

File: mlir/docs/Canonicalization.md
Comment (joker-eph) [inline mlir/docs/Canonicalization.md:111]: I don't understand this sentence (nor how it applies to what makes it a good pattern)

File: mlir/docs/Canonicalization.md
Comment (matthias-springer) [inline mlir/docs/Canonicalization.md:37]: What's the meaning of `more effective`? Either something has an effect or it doesn't. Do you mean `efficient`?

Also, I don't quite understand how that relates to the second sentence `Therefore, performance improvements are not necessary`.

File: mlir/docs/Canonicalization.md
Comment (joker-eph) [inline mlir/docs/Canonicalization.md:115]: Thanks for improving the doc! Can you split this out in a separate PR please?

File: mlir/docs/Canonicalization.md
Comment (matthias-springer) [inline mlir/docs/Canonicalization.md:59]: nit: `Canonicalization`. Also maybe `running time` is a better term because we are talking about the running time of the compiler. (Not the compile time of the C++ pattern implementation.)

File: mlir/docs/Canonicalization.md
Comment (matthias-springer) [inline mlir/docs/Canonicalization.md:107]: `The rewrite process always converges, because there is no canonicalization pattern that performs the inverse transformation.`

File: mlir/docs/Canonicalization.md
Comment (matthias-springer) [inline mlir/docs/Canonicalization.md:109]: Note quite sure what this part means `which is a clear lattice.` Maybe just drop it, "reducing the amount of computation" makes sense to me in this example.

File: mlir/docs/Canonicalization.md
Comment (matthias-springer) [inline mlir/docs/Canonicalization.md:110]: What exactly does `one-off` mean? Does it mean that this is a kind of canonicalization does is not related to any of the other existing canonicalizations? If so, why is that a bad thing?

Comment (stellaraccident) [reply mlir/docs/Canonicalization.md:110]: Good q. I think I've seen people using this two ways: a. Not part of a holistic approach to reducing all forms of the implicated ops, and b. Does not benefit from the overhead of being run in a fix point loop (often to say this is a lowering ala dialect conversion).

In this case, I believe the criticism was (a) but the author may have thought (b). 

It's not that it isn't related to any other canonicalization but that it isn't part of a holistic design.

I would probably split this into two:

- it is part of a holistic, consistent design for related forms.
- it is not a lowering that would be better maintained in a library of one off patterns that can be included in such pipelines.

File: mlir/include/mlir/Dialect/Utils/IndexingUtils.h
Comment (dcaballe) [reply mlir/include/mlir/Dialect/Utils/IndexingUtils.h:250]: Good point. My comment was motivated by the high proliferation of AffineMap and permutation utilities over the years (to the point that sometimes it's a challenge, even for people familiar with the code, to figure out if something exists already). However, I think this adding this one is justified as it's combined with other utilities that work on permutations.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
Comment (dcaballe) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1913]: nit: move ub to var + use pre-increment + remove curly braces for single statement `for` per coding guidelines

File: mlir/lib/Dialect/Utils/IndexingUtils.cpp
Comment (dcaballe) [inline mlir/lib/Dialect/Utils/IndexingUtils.cpp:260]: same

**Review by dcaballe — COMMENTED** [review]
The code LGTM. I would move the doc changes to a separate PR, as suggested.

**Comment by cxy-1993** [comment]
Thanks for the comments! Based on the comment's suggestion, the documentation part of this PR has been moved to https://github.com/llvm/llvm-project/pull/99753

**Comment by cxy-1993** [comment]
> The code LGTM. I would move the doc changes to a separate PR, as suggested.

Nice catch! I have modify the code according to your comment and move the doc to https://github.com/llvm/llvm-project/pull/99753 .

**Review by dcaballe — APPROVED** [review]
LGTM, thanks for bearing with us :)

**Pull request merged**
